lib/validate_params/validatable.rb

```diff
@@ -81,53 +70,57 @@ def perform_validate_params
+          elsif validation.type == Hash
             # Skip in case hash is configured and string is passed
-            next if hashlike?(params[key])
-            next if params.dig(key, validation[:field][key])
-
-            value = if validation[:options][:default].is_a?(Proc)
-                      validation[:options][:default].call
-                    else
-                      validation[:options][:default]
-                    end
+            next if params[validation.field].is_a?(String)
 
-            params[key] ||= {}
-            params[key][validation[:field][key]] = value
+            params[validation.field] ||= {}
+            apply_default_values(params[validation.field], sub_validation)
+          else
+            apply_default_values(params, sub_validation)
           end
-        else
-          value = if validation[:options][:default].is_a?(Proc)
-                    validation[:options][:default].call
-                  else
-                    validation[:options][:default]
-                  end
-
-          params[validation[:field]] ||= value
         end
+
+        return if validation.children.any?
+
+        options = validation.options.presence || {}
+        return unless options.key?(:default)
+
+        value = options[:default].is_a?(Proc) ? options[:default].call : options[:default]
+        params[validation.field] ||= value
       end
 
-      def cast_param_values(validations)
-        validations.each do |validation|
-          if validation[:field].is_a?(Hash)
-            validation[:field].each_key do |key|
-              next unless hashlike?(params[key])
+      def cast_param_values(params, validation)
+        return unless params
 
-              value = params.dig(key, validation[:field][key])
-              next if value.blank?
+        validation.children.each do |sub_validation|
+          if validation.type == Hash
+            # Skip in case hash is configured and string is passed
+            next if params[validation.field].is_a?(String)
 
-              params[key][validation[:field][key]] = Types.const_get(validation[:type].name).cast(value, **validation.fetch(:options, {}))
+            cast_param_values(params[validation.field], sub_validation)
+          elsif validation.type == Array
+            params[validation.field].each do |sub_params|
+              cast_param_values(sub_params, sub_validation)
             end
-          else
-            value = params[validation[:field]]
-            next if value.blank?
-
-            params[validation[:field]] = Types.const_get(validation[:type].name).cast(value, **validation.fetch(:options, {}))
           end
         end
+
+        return if validation.children.any?
+
+        value = params[validation.field]
+        return if value.blank?
+
+        options = validation.options.presence || {}
+        params[validation.field] = Types.const_get(validation.type.name).cast(value, **options)
       end
 
       def hashlike?(obj)
```

Comment:
Please check if we need this method, seems like it is not used right now.

Reply:
Good catch, I think we need to remove this method since it's no longer used.